src/pool/srv_metrics.c

```diff
@@ -37,59 +43,176 @@ ds_pool_metrics_init(void)
 }
 
+static void
+per_pool_metrics_lock(void)
+{
+	int rc = D_MUTEX_LOCK(&per_pool_lock);
+
+	if (unlikely(rc != 0))
+		D_ERROR("failed to lock per-pool metrics, "DF_RC"\n",
+			DP_RC(rc));
+}
+
+static void
+per_pool_metrics_unlock(void)
+{
+	int rc = D_MUTEX_UNLOCK(&per_pool_lock);
+
+	if (unlikely(rc != 0))
+		D_ERROR("failed to unlock per-pool metrics, "DF_RC"\n",
+			DP_RC(rc));
+}
+
+static void
+free_per_pool_metrics(struct per_pool_entry *entry)
+{
+	if (entry == NULL)
+		return;
+
+	d_list_del(&entry->link);
+	D_MUTEX_DESTROY(&entry->metrics.pm_lock);
+	D_FREE(entry);
+}
+
+static void
+close_pool_metrics_dir(const uuid_t pool_uuid)
+{
+	char	path[D_TM_MAX_NAME_LEN] = {0};
+	int	rc;
+
+	rc = ds_pool_metrics_get_path(pool_uuid, path, sizeof(path));
+	if (rc != 0) {
+		D_ERROR(DF_UUID ": unable to get pool metrics path, "DF_RC"\n",
+			DP_UUID(pool_uuid), DP_RC(rc));
+		return;
+	}
+
+	rc = d_tm_del_ephemeral_dir(path);
+	if (rc != 0) {
+		D_ERROR(DF_UUID ": unable to remove metrics dir for pool, "
+			DF_RC "\n", DP_UUID(pool_uuid), DP_RC(rc));
+		return;
+	}
+}
+
 /**
  * Finalizes the pool metrics
  */
 int
 ds_pool_metrics_fini(void)
 {
-	struct active_pool_metrics *cur = NULL;
-	struct active_pool_metrics *next = NULL;
+	struct per_pool_entry	*cur = NULL;
+	struct per_pool_entry	*next = NULL;
+	uuid_t			 uuid;
 
+	per_pool_metrics_lock();
 	d_list_for_each_entry_safe(cur, next, &per_pool_metrics, link) {
-		d_list_del(&cur->link);
-		D_FREE(cur);
+		uuid_copy(uuid, cur->metrics.pm_pool_uuid);
+		free_per_pool_metrics(cur);
+		close_pool_metrics_dir(uuid);
 	}
-	return 0;
+	per_pool_metrics_unlock();
+
+	D_DEBUG(DB_TRACE, "Finalized pool metrics\n");
+
+	return D_MUTEX_DESTROY(&per_pool_lock);
 }
 
 /**
- * Add metrics for a specific pool UUID.
+ * Create the metrics path for a specific pool UUID.
  *
- * \param[in]	pool_uuid	Pool UUID
+ * \param[in]	pool_uuid	UUID of the pool
+ * \param[out]	path		Path to the pool metrics
+ * \param[in]	path_len	Length of path array
+ *
+ * \return	0		Success
+ *		-DER_INVAL	Invalid inputs
  */
-void
-ds_pool_metrics_start(uuid_t pool_uuid)
+int
+ds_pool_metrics_get_path(const uuid_t pool_uuid, char *path, size_t path_len)
 {
-	struct active_pool_metrics	*metrics;
-	char				 uuid_str[DAOS_UUID_STR_SIZE];
-	char				 path[D_TM_MAX_NAME_LEN] = {0};
-	int				 rc;
+	char uuid_str[DAOS_UUID_STR_SIZE];
 
 	if (!daos_uuid_valid(pool_uuid)) {
 		D_ERROR(DF_UUID ": invalid uuid\n", DP_UUID(pool_uuid));
-		return;
+		return -DER_INVAL;
 	}
+	uuid_unparse(pool_uuid, uuid_str);
+
+	snprintf(path, path_len, POOL_METRICS_DIR_FMT, uuid_str);
+	path[path_len - 1] = '\0';
+	return 0;
+}
+
+static int
+new_per_pool_metrics(const uuid_t pool_uuid, const char *path)
+{
+	struct per_pool_entry	*entry;
+	int			 rc;
+
+	D_ALLOC_PTR(entry);
+	if (entry == NULL) {
+		D_ERROR(DF_UUID ": failed to allocate metrics struct\n",
+			DP_UUID(pool_uuid));
+		return -DER_NOMEM;
+	}
+
+	D_MUTEX_INIT(&entry->metrics.pm_lock, NULL);
```

Comment:
We cannot use mutex on the server side and have to use ABT mutexes instead.

Reply:
Ah, good catch. Will fix.